gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,215 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::primitives::ReservationId;
+use crate::{errors::Result, exec, prelude::*};
+
+struct Reservation {
+    id: ReservationId,
+    amount: u64,
+    valid_until: u32,
+}
+
+/// Result of [`Reservations::try_take_reservation`](Reservations::try_take_reservation)
+pub enum TakeReservationResult {
```

Comment:
Can't we just go with `-> Option<Reservation>` for api methods?
And reservation method would has get fns + api to send message with it